# Importing TensorFlow

In [ ]:
import tensorflow as tf
print("TensorFlow version: {}".format(tf.__version__))
print("Eager execution is: {}".format(tf.executing_eagerly()))
print("Keras version: {}".format(tf.keras.__version__))


TensorFlow version: 2.9.2
Eager execution is: True
Keras version: 2.9.0


#Coding style convention for TensorFlow
The Google Python Style Guide, which can be found at https://github.com/google/styleguide/blob/gh-pages/pyguide.md

#Using eager execution
Find out whether a CPU or GPU is in use

In [ ]:
var = tf.Variable([3, 3])
if tf.test.is_gpu_available():
 print('Running on GPU')
 print('GPU #0?')
 print(var.device.endswith('GPU:0'))
else:
 print('Running on CPU')

Instructions for updating:
Use `tf.config.list_physical_devices('GPU')` instead.


Running on CPU


In [ ]:
t0 = 24 # python variable
t1 = tf.Variable(42) # rank 0 tensor
t2 = tf.Variable([ [ [0., 1., 2.], [3., 4., 5.] ], [ [6., 7., 8.], [9., 10., 11.] ] ]) #rank 3 tensor
t0, t1, t2


(24,
 <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=42>,
 <tf.Variable 'Variable:0' shape=(2, 2, 3) dtype=float32, numpy=
 array([[[ 0.,  1.,  2.],
         [ 3.,  4.,  5.]],
 
        [[ 6.,  7.,  8.],
         [ 9., 10., 11.]]], dtype=float32)>)

##Declaring TensorFlow constants

In [ ]:
v = tf.constant(42)
v

<tf.Tensor: shape=(), dtype=int32, numpy=42>

In [ ]:
v.numpy()

42

In [ ]:
var = tf.constant(1, dtype = tf.int64)
var

<tf.Tensor: shape=(), dtype=int64, numpy=1>

##Shaping a tensor

In [ ]:
t2= tf.Variable([ [ [0., 1., 2.], [3., 4., 5.] ], [ [6., 7., 8.], [9., 10., 11.] ] ]) # tensor variable
print(t2.shape)


(2, 2, 3)


In [ ]:
r1 = tf.reshape(t2,[2,6]) # 2 rows 6 cols
r2 = tf.reshape(t2,[1,12]) # 1 rows 12 cols
r1

<tf.Tensor: shape=(2, 6), dtype=float32, numpy=
array([[ 0.,  1.,  2.,  3.,  4.,  5.],
       [ 6.,  7.,  8.,  9., 10., 11.]], dtype=float32)>

In [ ]:
r2 = tf.reshape(t2,[1,12]) # 1 row 12 columns
r2

<tf.Tensor: shape=(1, 12), dtype=float32, numpy=
array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11.]],
      dtype=float32)>

##Ranking (dimensions) of a tensor

In [ ]:
tf.rank(t2)

<tf.Tensor: shape=(), dtype=int32, numpy=3>

In [ ]:
t3 = t2[1, 0, 2] # slice 1, row 0, column 2
t3

<tf.Tensor: shape=(), dtype=float32, numpy=8.0>

#Casting a tensor to a NumPy/Python variable


In [ ]:
print(t2.numpy())


[[[ 0.  1.  2.]
  [ 3.  4.  5.]]

 [[ 6.  7.  8.]
  [ 9. 10. 11.]]]


In [ ]:
print(t2[1, 0, 2].numpy())

8.0


##Finding the size (number of elements) of a tensor

In [ ]:
tf.size(input=t2).numpy()

12

#Finding the datatype of a tensor

In [ ]:
t3.dtype

tf.float32

#Specifying element-wise primitive tensor operations
Element-wise primitive tensor operations are specified using, as you would expect, the overloaded operators +, -, *, and /, as here:


In [ ]:
t2*t2

<tf.Tensor: shape=(2, 2, 3), dtype=float32, numpy=
array([[[  0.,   1.,   4.],
        [  9.,  16.,  25.]],

       [[ 36.,  49.,  64.],
        [ 81., 100., 121.]]], dtype=float32)>

In [ ]:
t4 = t2*4
print(t4)

tf.Tensor(
[[[ 0.  4.  8.]
  [12. 16. 20.]]

 [[24. 28. 32.]
  [36. 40. 44.]]], shape=(2, 2, 3), dtype=float32)


#Transposing TensorFlow and matrix multiplication

In [ ]:
u = tf.constant([[3,4,3]])
v = tf.constant([[1,2,1]])
tf.matmul(u, tf.transpose(a=v))


<tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[14]], dtype=int32)>

There is a complete list of these operations at https://www.tensorflow.org/api_docs/python/tf/math?hl=en

#Casting a tensor to another (tensor) datatype

In [ ]:
i = tf.cast(t1, dtype=tf.int32) # 42
i

<tf.Variable 'Variable:0' shape=() dtype=int32, numpy=42>

In [ ]:
j = tf.cast(tf.constant(4.9), dtype=tf.int32) # 4
j

<tf.Tensor: shape=(), dtype=int32, numpy=4>

#Declaring ragged tensors
A ragged tensor is a tensor with one or more ragged dimensions. Ragged dimensions are
dimensions that have slices that may have different lengths.

In [ ]:
ragged =tf.ragged.constant([[5, 2, 6, 1], [], [4, 10, 7], [8], [6,7]])
print(ragged)
print(ragged[0,:])
print(ragged[1,:])
print(ragged[2,:])
print(ragged[3,:])
print(ragged[4,:])


<tf.RaggedTensor [[5, 2, 6, 1], [], [4, 10, 7], [8], [6, 7]]>
tf.Tensor([5 2 6 1], shape=(4,), dtype=int32)
tf.Tensor([], shape=(0,), dtype=int32)
tf.Tensor([ 4 10  7], shape=(3,), dtype=int32)
tf.Tensor([8], shape=(1,), dtype=int32)
tf.Tensor([6 7], shape=(2,), dtype=int32)


#Finding the squared difference between two tensors


In [ ]:
x = [1,3,5,7,11]
y = 5
s = tf.math.squared_difference(x,y)
s

<tf.Tensor: shape=(5,), dtype=int32, numpy=array([16,  4,  0,  4, 36], dtype=int32)>

#Finding a mean

In [ ]:
#tf.reduce_mean() 
# equivalent to np.mean
#tf.reduce_mean(input_tensor, axis=None, keepdims=None, name=None)

numbers = tf.constant([[4., 5.], [7., 3.]])
tf.reduce_mean(input_tensor=numbers)


<tf.Tensor: shape=(), dtype=float32, numpy=4.75>

In [ ]:
tf.reduce_mean(input_tensor=numbers, axis=0) # [ (4. + 7. )/2 , (5. + 3.)/2] = [5.5, 4.]


<tf.Tensor: shape=(2,), dtype=float32, numpy=array([5.5, 4. ], dtype=float32)>

In [ ]:
# axis=None finding the mean across all axes
# axis=0 finding the mean across columns
# axis=1 finding themean across rows

#Generating tensors filled with random values

In [ ]:
#tf. random.normal(shape, mean = 0, stddev =2, dtype=tf.float32, seed=None, name=None)
tf.random.normal(shape = (3,2), mean=10, stddev=2, dtype=tf.float32, seed=None, name=None)
ran = tf.random.normal(shape = (3,2), mean=10.0, stddev=2.0)
print(ran)

tf.Tensor(
[[11.051319  13.0052395]
 [ 9.547969   5.199543 ]
 [11.651008  12.191878 ]], shape=(3, 2), dtype=float32)


In [ ]:
tf.random.uniform(shape = (2,4), minval=0, maxval=None, dtype=tf.float32, seed=None, name=None)

<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0.11415112, 0.9734775 , 0.12194204, 0.6146188 ],
       [0.16630757, 0.77495646, 0.89908445, 0.65204227]], dtype=float32)>

#Finding the indices of the largest and smallest element


In [ ]:
#tf.argmax(input, axis=None, name=None, output_type=tf.int64 )
#tf.argmin(input, axis=None, name=None, output_type=tf.int64 )
# 1-D tensor
t5 = tf.constant([2, 11, 5, 42, 7, 19, -6, -11, 29])
print(t5)
i = tf.argmax(input=t5)
print('index of max; ', i)
print('Max element: ',t5[i].numpy())

tf.Tensor([  2  11   5  42   7  19  -6 -11  29], shape=(9,), dtype=int32)
index of max;  tf.Tensor(3, shape=(), dtype=int64)
Max element:  42


#Using tf.function
tf.function is a function that will take a Python function and return a TensorFlow graph.
Its signature is as follows:
`tf.function(
 func=None,
 input_signature=None,
 autograph=True,
 experimental_autograph_options=None
)`


In [ ]:
def f1(x, y):
 return tf.reduce_mean(input_tensor=tf.multiply(x ** 2, 5) + y**2)
f2 = tf.function(f1)
x = tf.constant([4., -5.])
y = tf.constant([2., 3.])
# f1 and f2 return the same value, but f2 executes as a TensorFlow graph
assert f1(x,y).numpy() == f2(x,y).numpy()


#The features of Keras

In [ ]:
import tensorflow as tf
print(tf.keras.__version__)

2.9.0


#The Keras backend
Due to its model-level library structure, Keras may have different tensor manipulation engines that handle low-level operations, such as convolutions, tensor products, and the like. These engines are called backends.

In [ ]:
from tensorflow.keras import backend as K
const = K.constant([[42,24],[11,99]], dtype=tf.float16, shape=[2,2])
const


<tf.Tensor: shape=(2, 2), dtype=float16, numpy=
array([[42., 24.],
       [11., 99.]], dtype=float16)>

#Keras data types
Keras data types (dtypes) are the same as TensorFlow Python data types

tf.float16  
tf.float32  
tf.float64  
tf.int8  
tf.int16  
tf.int32  
tf.int64  
tf.uint8  
tf.string  
tf.bool Boolean  
tf.complex64   
tf.complex128  
tf.qint8  
tf.qint32  
tf.quint8 




#Keras models
Keras is based on the concept of a neural network model. The predominant model is called a Sequence, being a linear stack of layers.
#The Keras Sequential model
1 - Add layers  
2 - Compile it  
3 - Fit the model to the data  
4 - Evaluate your model to establish its accuracy, loss, and other metrics.   
5 - Use it to make predictions on new data  

#There are two ways to create a Sequential model.

`Flatten` takes the input of 28 x 28 (that is, 2D) pixel images and produces a 784
(that is, 1D) vector because the next (dense) layer is one-dimensional.

`Dense` is a fully connected layer, meaning all its neurons are connected to every neuron in the previous and next layers. The following example has 512 neurons, and its inputs are passed through a ReLU (non-linear) activation function. 
  
`Dropout` randomly turns off a fraction (in this case, 0.2) of neurons in the
previous layer. This is done to prevent any particular neuron becoming too
specialized and causing overfitting of the model to the data, thus impacting on the accuracy metric of the model on the test data.

The final Dense layer has a special activation function called `softmax`, which
assigns probabilities to each of the possible 10 output units:


In [ ]:
mnist = tf.keras.datasets.mnist
(train_x,train_y), (test_x, test_y) = mnist.load_data()

11490434/11490434 [==============================] - 0s 0us/step


The `epochs` variable stores the number of times we are going to present the data to the model.

In [ ]:
epochs=10
batch_size = 32 # 32 is default in fit method but specify anyway

Normalize all the data points (x) to be in the float range zero to one, and of
the float32 type. Also, cast the labels (y) to int64, as required

In [ ]:
train_x, test_x = tf.cast(train_x/255.0, tf.float32), tf.cast(test_x/255.0,
tf.float32)
train_y, test_y = tf.cast(train_y,tf.int64),tf.cast(test_y,tf.int64)

In [ ]:
#The first way to create a Sequential model
model1 = tf.keras.models.Sequential([
 tf.keras.layers.Flatten(),
 tf.keras.layers.Dense(512,activation=tf.nn.relu),
 tf.keras.layers.Dropout(0.2),
 tf.keras.layers.Dense(10,activation=tf.nn.softmax)
])

In [ ]:
# Compile our model
optimiser = tf.keras.optimizers.Adam()
model1.compile (optimizer= optimiser,
loss='sparse_categorical_crossentropy', metrics = ['accuracy'])

In [ ]:
# To train our model, we use the fit method  
model1.fit(train_x, train_y, batch_size=batch_size, epochs=epochs)

Epoch 1/10
1875/1875 [==============================] - 9s 4ms/step - loss: 0.2197 - accuracy: 0.9356
Epoch 2/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0973 - accuracy: 0.9703
Epoch 3/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0699 - accuracy: 0.9784
Epoch 4/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0537 - accuracy: 0.9823
Epoch 5/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0446 - accuracy: 0.9856
Epoch 6/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0365 - accuracy: 0.9878
Epoch 7/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0309 - accuracy: 0.9903
Epoch 8/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0283 - accuracy: 0.9905
Epoch 9/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0250 - accuracy: 0.9921
Epoch 10/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.02

In [ ]:
 model1.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 784)               0         
                                                                 
 dense (Dense)               (None, 512)               401920    
                                                                 
 dropout (Dropout)           (None, 512)               0         
                                                                 
 dense_1 (Dense)             (None, 10)                5130      
                                                                 
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# finally, we can check our trained model for accuracy using the evaluate method:
model1.evaluate(test_x, test_y)


313/313 [==============================] - 1s 2ms/step - loss: 0.0769 - accuracy: 0.9814


[0.07689265161752701, 0.9814000129699707]

#The second way to create a Sequential model


In [ ]:
model2 = tf.keras.models.Sequential();
model2.add(tf.keras.layers.Flatten())
model2.add(tf.keras.layers.Dense(512, activation='relu'))
model2.add(tf.keras.layers.Dropout(0.2))
model2.add(tf.keras.layers.Dense(10,activation=tf.nn.softmax))
model2.compile (optimizer=optimiser, loss='sparse_categorical_crossentropy', 
                metrics = ['accuracy'])


In [ ]:
 model2.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_3 (Flatten)         (None, 784)               0         
                                                                 
 dense_6 (Dense)             (None, 512)               401920    
                                                                 
 dropout_3 (Dropout)         (None, 512)               0         
                                                                 
 dense_7 (Dense)             (None, 10)                5130      
                                                                 
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model2.fit(train_x, train_y, batch_size=batch_size, epochs=epochs)
model2.evaluate(test_x, test_y)

Epoch 1/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1854 - accuracy: 0.9445
Epoch 2/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0873 - accuracy: 0.9731
Epoch 3/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0619 - accuracy: 0.9807
Epoch 4/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0491 - accuracy: 0.9844
Epoch 5/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0382 - accuracy: 0.9875
Epoch 6/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0339 - accuracy: 0.9892
Epoch 7/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0305 - accuracy: 0.9902
Epoch 8/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.0265 - accuracy: 0.9912
Epoch 9/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0239 - accuracy: 0.9920
Epoch 10/10
313/313 [==============================] - 1s 2ms/step - loss: 0.0697

[0.06972582638263702, 0.9843999743461609]

#The Keras functional API

In [ ]:
#The setup code is the same as previously demonstrated:
#import tensorflow as tf
mnist = tf.keras.datasets.mnist
(train_x,train_y), (test_x, test_y) = mnist.load_data()
train_x, test_x = train_x/255.0, test_x/255.0
epochs=10



In [ ]:
inputs = tf.keras.Input(shape=(28,28)) # Returns a 'placeholder' tensor
x = tf.keras.layers.Flatten()(inputs)
x = tf.keras.layers.Dense(512, activation='relu',name='d1')(x)
x = tf.keras.layers.Dropout(0.2)(x)
predictions = tf.keras.layers.Dense(10,activation=tf.nn.softmax,
name='d2')(x)
model3 = tf.keras.Model(inputs=inputs, outputs=predictions)

In [ ]:
 model3.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_2 (InputLayer)        [(None, 28, 28)]          0         
                                                                 
 flatten_5 (Flatten)         (None, 784)               0         
                                                                 
 d1 (Dense)                  (None, 512)               401920    
                                                                 
 dropout_4 (Dropout)         (None, 512)               0         
                                                                 
 d2 (Dense)                  (None, 10)                5130      
                                                                 
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


In [ ]:
optimiser = tf.keras.optimizers.Adam()
model3.compile (optimizer= optimiser,
loss='sparse_categorical_crossentropy', metrics = ['accuracy'])
model3.fit(train_x, train_y, batch_size=32, epochs=epochs)
model3.evaluate(test_x, test_y)


Epoch 1/10
1875/1875 [==============================] - 9s 4ms/step - loss: 0.2175 - accuracy: 0.9355
Epoch 2/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0957 - accuracy: 0.9710
Epoch 3/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0699 - accuracy: 0.9783
Epoch 4/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0541 - accuracy: 0.9829
Epoch 5/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0428 - accuracy: 0.9861
Epoch 6/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0353 - accuracy: 0.9890
Epoch 7/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0320 - accuracy: 0.9890
Epoch 8/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0270 - accuracy: 0.9910
Epoch 9/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0257 - accuracy: 0.9908
Epoch 10/10
313/313 [==============================] - 1s 2ms/step - loss: 0.0754 

[0.07535390555858612, 0.9799000024795532]

In [ ]:
#Full example 
import tensorflow as tf
mnist = tf.keras.datasets.mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test, y_test)

Epoch 1/5
1875/1875 [==============================] - 5s 2ms/step - loss: 0.2998 - accuracy: 0.9129
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1450 - accuracy: 0.9570
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1067 - accuracy: 0.9672
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0880 - accuracy: 0.9729
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0725 - accuracy: 0.9766


[0.07250205427408218, 0.9765999913215637]

In [ ]:
class MyModel(tf.keras.Model):
  def __init__(self, num_classes=10):
    super(MyModel, self).__init__()
    # Define your layers here.
    inputs = tf.keras.Input(shape=(28,28)) # Returns a placeholder tensor
    self.x0 = tf.keras.layers.Flatten()
    self.x1 = tf.keras.layers.Dense(512, activation='relu',name='d1')
    self.x2 = tf.keras.layers.Dropout(0.2)
    self.predictions = tf.keras.layers.Dense(10,activation=tf.nn.softmax,name='d2')
  def call(self, inputs):
  # This is where to define your forward pass
  # using the layers previously defined in `__init__`
    x = self.x0(inputs)
    x = self.x1(x)
    x = self.x2(x)
    return self.predictions(x)

model4 = MyModel()
batch_size = 32
#steps_per_epoch = len(train_x.numpy())//batch_size
#print(steps_per_epoch)
model4.compile (optimizer= optimiser, loss='sparse_categorical_crossentropy',
                metrics = ['accuracy'])
model4.fit(train_x, train_y, batch_size=batch_size, epochs=epochs)
model4.evaluate(test_x, test_y)



Epoch 1/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1868 - accuracy: 0.9442
Epoch 2/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0864 - accuracy: 0.9731
Epoch 3/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0629 - accuracy: 0.9802
Epoch 4/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0484 - accuracy: 0.9844
Epoch 5/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.0402 - accuracy: 0.9868
Epoch 6/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0332 - accuracy: 0.9891
Epoch 7/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0272 - accuracy: 0.9910
Epoch 8/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0264 - accuracy: 0.9911
Epoch 9/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0228 - accuracy: 0.9924
Epoch 10/10
313/313 [==============================] - 1s 2ms/step - loss: 0.0709

[0.07089351862668991, 0.9832000136375427]

#Saving and loading Keras models

In [ ]:
#Saving a model 
model.save('./model_name.h5')

#Loading a saved model
from tensorflow.keras.models import load_model
new_model = load_model('./model_name.h5')

#<font color=red>To do</font>
Train a deep MLP on the MNIST dataset (you can load it using `tf.keras.​data⁠sets.mnist.load_data()`. See if you can get over 98% accuracy by manually tuning the hyperparameters. Try searching for the optimal learning rate (by growing the learning rate exponentially, plotting the loss, and finding the point where the loss shoots up). Next, try tuning the hyperparameters using Keras Tuner with all the bells and whistles—save checkpoints, use early stopping, and plot learning curves using TensorBoard.

#<font color=red>To do (Extra) </font> 
## Deep neural network on the CIFAR10 image dataset:

### a.
Build a DNN with 20 hidden layers of 100 neurons each (that’s too many, but it’s the point of this exercise). Use He initialization and the Swish activation function.
### b.
Using Nadam optimization and early stopping, train the network on the CIFAR10 dataset. You can load it with `tf.keras.datasets.cifar10.load_​data()`. The dataset is composed of 60,000 32 × 32–pixel color images (50,000 for training, 10,000 for testing) with 10 classes, so you’ll need a softmax output layer with 10 neurons. Remember to search for the right learning rate each time you change the model’s architecture or hyperparameters.
### c.
Now try adding batch normalization and compare the learning curves: is it converging faster than before? Does it produce a better model? How does it affect training speed?
### d.
Try replacing batch normalization with SELU, and make the necessaryadjustments to ensure the network self-normalizes (i.e., standardize the input features, use LeCun normal initialization, make sure the DNN contains only a sequence of dense layers, etc.).
### e.
Try regularizing the model with alpha dropout. Then, without retraining your model, see if you can achieve better accuracy using MC dropout.
### f.
Retrain your model using 1cycle scheduling and see if it improves training speed and model accuracy.